Enforce single org for series key reads

stuartcarnie opened the pull request:
_Briefly describe your proposed changes:_

These changes modify the parameters to a series cursor request and render the intent explicitly; series keys returned by the cursor must never cross organization boundaries.

* `SeriesCursorRequest` takes a single `Name` parameter, specifying the 16-byte tsdb encoded org and bucket ID
* `seriesCursor` ensures the first 8 bytes of the measurement name returned from the series index matches the current organization ID

**Blocked by #12010**

e-dard commented on storage/series_cursor.go:
```diff
@@ -98,6 +103,9 @@ func (cur *seriesCursor) Next() (*SeriesCursorRow, error) {
 		}
 
 		cur.row.Name, cur.row.Tags = tsdb.ParseSeriesKey(cur.keys[cur.ofs])
+		if !bytes.HasPrefix(cur.row.Name, cur.orgID) {
```
I think we should either change the `orgID` type to something stronger, or if we leave `orgID` as a `[]byte` then maybe `len(cur.row.Name) < 8 || !bytes.Equal(cur.row.Name[:8], cur.orgID)`.

stuartcarnie replied:
`bytes.HasPrefix` is implemented as 

```go
len(name) >= len(orgID) && Equal(name[0:len(prefix)], orgID)
```

So by adding an assertion in `newSeriesCursor`, such as `len(orgID) == influxdb.IDBytesLen`, we could ensure this invariant.

e-dard commented on storage/series_cursor_test.go:
```diff
@@ -0,0 +1,38 @@
+			makeKey(0x0f0f, 0xb0b0),
+			makeKey(0xffff, 0xb0b0),
+		},
+		orgID: nm[:8],
```
Can we change `[:8]` to `[:influxdb.OrgIDLength]` or something? I just want to make it simpler to change the id scheme if we need to.

stuartcarnie replied:
Yeah, sounds good; I'll use that constant in the `newSeriesCursor` to validate too

stuartcarnie commented on storage/series_cursor.go:
```diff
@@ -11,20 +12,28 @@ import (
 	index *tsi1.Index
+	sfile *tsdb.SeriesFile
 	mitr  tsdb.MeasurementIterator
+	orgID []byte
```
We could assert `len(orgID) == influxdb.IDBytesLen` or similar in `newSeriesCursor` to strengthen this precondition

stuartcarnie commented:
@e-dard, cleaned up the code a bit more:

* replaced `orgID` with `name [influxdb.IDLength]byte` and use that to test the prefix with `name[:influxdb.OrgIDLength]`
* removed all the measurement iterator stuff, as it was redundant

That end-to-end test you shared does exercise this new code and passes, however, there is no easy way to test the negative case, besides explicitly adding ways to break the invariants in the series and index logic.
